### DISCUSSION TITLE
SQLModel + Pydantic v1 vs SQLModel + Pydantic v2 - different assignment behaviour

### DISCUSSION THREAD

**Discussion by ta-soft** [fastapi/sqlmodel#824] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from sqlmodel import SQLModel, Field
from datetime import date
from typing import Optional


class A(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    dob: date

class B(BaseModel):
    dob: date

a=A(dob='1980-01-01')
print(a)
b=B(dob='1980-01-01')
print(b)
```

### Description

The code above prints
```
id=None dob=datetime.date(1980, 1, 1)
dob=datetime.date(1980, 1, 1)
```
if executed with pydantic v1 in the mix (1)*.

But it prints
```
dob='1980-01-01' id=None
dob=datetime.date(1980, 1, 1)
```
With pydantic v2 in the mix (2)*. 

This behaviour essentially breaks a fastapi application I am migrating to pedantic v2, because I assume, for example, that date attributes within objects passed in the body of post methods are already converted (originally passed as strings).

I couldn't find any recommendation related to this. I realize I could set table=False on the SQLModel object, but if I do that other behaviours change, related to validations for example.

Am I missing something?

(1) pyproject dependencies
``` 

python = "^3.10"
fastapi = "^0.110.0"
sqlmodel = "^0.0.16"
uvicorn = "^0.27.1"
pydantic = "1.10.14"
```

(2) pyproject deps
``` 
python = "^3.10"
fastapi = "^0.110.0"
sqlmodel = "^0.0.16"
uvicorn = "^0.27.1"
pydantic = "2.6.3"
```

### Operating System

Linux

### SQLModel Version

0.0.16

### Python Version

3.10

**Comment by ta-soft** [comment]
note: adding 

```
model_config = ConfigDict(validate_assignment=True)
```
to my models is not really an option, because that will make the code also start validating objects as soon as they are created, not allowing different attributes to be populated in phases (which is something I suspect a lot of people do frequently)

**Comment by Viicos** [comment]
Is your issue about the fact that in Pydantic v2, the `dob` attribute isn't parsed as a `datetime.date` object?

**Reply by ta-soft** [reply]
Generally speaking v2 + sqlmodel appear to not coerce/convert types where possible - and that is my issue

**Comment by ta-soft** [comment]
Any updates or thoughts on this one?

**Comment by YuriiMotov** [comment, marked as answer]
SQLModels with `table=True` bypass validation, you shouldn't use models with `table=True` for input validation.

Create `ABase` model, move common fields there and use it for validation:

```py
from sqlmodel import SQLModel, Field
from datetime import date
from typing import Optional


class ABase(SQLModel):
    id: Optional[int] = Field(default=None, primary_key=True)
    dob: date


class A(ABase, table=True):
    pass


a = ABase(dob="1980-01-01")
print(a)  # id=None dob=datetime.date(1980, 1, 1)
```
